### PR TITLE
Update spdk gRPC for online rebuilding

### PR DESCRIPTION
**Pull request by shuo-wu**
Longhorn/longhorn#7199

#### Which issue(s) this PR fixes:

Issue #

### CODE REVIEW

File: protobuf/spdkrpc/spdk.proto [hunk 1]
```diff
@@ -15,15 +15,18 @@ service SPDKService {
...
     rpc ReplicaWatch(google.protobuf.Empty) returns (stream google.protobuf.Empty) {}
-    rpc ReplicaRebuildingSrcStart(ReplicaRebuildingSrcStartRequest) returns (google.protobuf.Empty) {}
-    rpc ReplicaRebuildingSrcFinish(ReplicaRebuildingSrcFinishRequest) returns (google.protobuf.Empty) {}
-    rpc ReplicaRebuildingSrcAttach(ReplicaRebuildingSrcAttachRequest) returns (google.protobuf.Empty) {}
-    rpc ReplicaRebuildingSrcDetach(ReplicaRebuildingSrcDetachRequest) returns (google.protobuf.Empty) {}
-    rpc ReplicaSnapshotShallowCopy(ReplicaSnapshotShallowCopyRequest) returns (google.protobuf.Empty);
-    rpc ReplicaRebuildingDstStart(ReplicaRebuildingDstStartRequest) returns (ReplicaRebuildingDstStartResponse) {}
-    rpc ReplicaRebuildingDstFinish(ReplicaRebuildingDstFinishRequest) returns (google.protobuf.Empty) {}
+    rpc ReplicaRebuildingSrcStart(ReplicaRebuildingSrcStartRequest) returns (ReplicaRebuildingSrcStartResponse) {};
+    rpc ReplicaRebuildingSrcFinish(ReplicaRebuildingSrcFinishRequest) returns (google.protobuf.Empty) {};
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:19]: Don't we need any info in the response?

File: protobuf/spdkrpc/spdk.proto [hunk 2]
```diff
@@ -15,15 +15,18 @@ service SPDKService {
...
+    rpc ReplicaRebuildingSrcAttach(ReplicaRebuildingSrcAttachRequest) returns (google.protobuf.Empty) {};
+    rpc ReplicaRebuildingSrcDetach(ReplicaRebuildingSrcDetachRequest) returns (google.protobuf.Empty) {};
+    rpc ReplicaRebuildingSrcShallowCopyStart(ReplicaRebuildingSrcShallowCopyStartRequest) returns (ReplicaRebuildingSrcShallowCopyStartResponse);
+    rpc ReplicaRebuildingSrcShallowCopyCheck(ReplicaRebuildingSrcShallowCopyCheckRequest) returns (ReplicaRebuildingSrcShallowCopyCheckResponse) {};
+    rpc ReplicaRebuildingDstStart(ReplicaRebuildingDstStartRequest) returns (ReplicaRebuildingDstStartResponse) {};
+    rpc ReplicaRebuildingDstFinish(ReplicaRebuildingDstFinishRequest) returns (google.protobuf.Empty) {};
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:25]: Same as above

File: protobuf/spdkrpc/spdk.proto [hunk 3]
```diff
@@ -15,15 +15,18 @@ service SPDKService {
...
+    rpc ReplicaRebuildingDstShallowCopyStart(ReplicaRebuildingDstShallowCopyStartRequest) returns (google.protobuf.Empty);
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:26]: Does this RPC start a shallow copy? If yes, shouldn't the response contain the operation_id?

**Review by DamiaSan — COMMENTED** [review]
LGTM, just some questions

File: protobuf/spdkrpc/spdk.proto [hunk 4]
```diff
@@ -140,23 +147,68 @@ message ReplicaRebuildingSrcDetachRequest {
...
+    string error_msg = 4;
+}
+
 message ReplicaRebuildingDstStartRequest {
     string name = 1;
-    bool expose_required = 2;
+    string src_replica_name = 2;
+    string src_replica_address = 3;
+    string external_snapshot_name = 4;
+    string external_snapshot_lvol_address = 5;
```
Comment (derekbit) [inline protobuf/spdkrpc/spdk.proto:178]: Is `external_snapshot_address` better?

File: protobuf/spdkrpc/spdk.proto [hunk 5]
```diff
@@ -121,7 +123,12 @@ enum ReplicaMode {
 message ReplicaRebuildingSrcStartRequest {
     string name = 1;
     string dst_replica_name = 2;
-    string dst_rebuilding_lvol_address = 3;
+    string dst_replica_address = 3;
+    string exposed_snapshot_name = 4;
+}
+
+message ReplicaRebuildingSrcStartResponse {
+    string exposed_snapshot_lvol_address = 1;
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:131]: The same as Derek's comment below, is `exposed_snapshot_address` better?

File: protobuf/spdkrpc/spdk.proto [hunk 6]
```diff
@@ -140,23 +147,68 @@ message ReplicaRebuildingSrcDetachRequest {
...
     string name = 1;
     string snapshot_name = 2;
 }
 
+message ReplicaRebuildingSrcShallowCopyStartResponse {
+    uint32 shallow_copy_op_id = 1;
+}
+
+message ReplicaRebuildingSrcShallowCopyCheckRequest {
+    string name = 1;
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:160]: Is `name` source replica's name right?

File: protobuf/spdkrpc/spdk.proto [hunk 7]
```diff
@@ -140,23 +147,68 @@ message ReplicaRebuildingSrcDetachRequest {
...
+
+message ReplicaRebuildingSrcShallowCopyCheckResponse {
+    string state = 1;
+    uint64 copied_clusters = 2;
+    uint64 total_clusters = 3;
+    string error_msg = 4;
+}
+
 message ReplicaRebuildingDstStartRequest {
     string name = 1;
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:174]: Is `name` destination replica's name right?

File: protobuf/spdkrpc/spdk.proto [hunk 8]
```diff
@@ -140,23 +147,68 @@ message ReplicaRebuildingSrcDetachRequest {
...
+    repeated Lvol rebuilding_snapshot_list = 6;
 }
 
 message ReplicaRebuildingDstStartResponse {
-    string address = 1;
+    string dst_head_lvol_address = 1;
 }
 
 message ReplicaRebuildingDstFinishRequest {
     string name = 1;
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:187]: Is `name` destination replica's name right?

File: protobuf/spdkrpc/spdk.proto [hunk 9]
```diff
@@ -140,23 +147,68 @@ message ReplicaRebuildingSrcDetachRequest {
...
-    bool unexpose_required = 2;
+}
+
+message ReplicaRebuildingDstShallowCopyStartRequest {
+    string name = 1;
+    string snapshot_name = 2;
+}
+
+message ReplicaRebuildingDstShallowCopyCheckRequest {
+    string name = 1;
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:196]: Is `name` destination replica's name right?

File: protobuf/spdkrpc/spdk.proto [hunk 10]
```diff
@@ -140,23 +147,68 @@ message ReplicaRebuildingSrcDetachRequest {
...
+    string dst_head_lvol_address = 1;
 }
 
 message ReplicaRebuildingDstFinishRequest {
     string name = 1;
-    bool unexpose_required = 2;
+}
+
+message ReplicaRebuildingDstShallowCopyStartRequest {
+    string name = 1;
```
Comment (DamiaSan) [inline protobuf/spdkrpc/spdk.proto:191]: Is `name` destination replica's name right?

File: protobuf/spdkrpc/spdk.proto (hunk 6 above)
Comment (shuo-wu) [reply protobuf/spdkrpc/spdk.proto:160]: Yes. This is always the receiver's name. Based on the API naming, the receiver is Rebuilding Src Replica.
Similar below.

**Review by derekbit — APPROVED** [review]

**Pull request merged**